ArmPkg/Drivers/SmmuDxe/IoMmu.c

```diff
@@ -0,0 +1,523 @@
+                  );
+  if (!EFI_ERROR (Status)) {
+    *HostAddress = (VOID *)(UINTN)PhysicalAddress;
+  }
+
+  return Status;
+}
+
+/**
+  Set the R/W attributes for a device handle.
+
+  @param [in]  This          Pointer to the IOMMU protocol instance.
+  @param [in]  DeviceHandle  The device handle to set attributes for.
+  @param [in]  Mapping       The mapping to set attributes for.
+  @param [in]  IoMmuAccess   The IOMMU access attributes for R/W.
+
+  @retval EFI_SUCCESS        Success.
+**/
+EFI_STATUS
+EFIAPI
+IoMmuSetAttribute (
+  IN EDKII_IOMMU_PROTOCOL  *This,
+  IN EFI_HANDLE            DeviceHandle,
+  IN VOID                  *Mapping,
+  IN UINT64                IoMmuAccess
+  )
+{
+  EFI_STATUS      Status;
+  IOMMU_MAP_INFO  *MapInfo;
+
+  if (Mapping == NULL) {
+    return EFI_SUCCESS;
+  }
+
+  MapInfo = (IOMMU_MAP_INFO *)Mapping;
+
+  Status = UpdatePageTable (
+             mSmmu->PageTableRoot,
+             MapInfo->PA,
+             MapInfo->NumberOfBytes,
+             PAGE_TABLE_READ_WRITE_FROM_IOMMU_ACCESS (IoMmuAccess),
+             FALSE,
+             TRUE
+             );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: UpdatePageTable failed.\n", __func__));
+    return Status;
+  }
+
+  return Status;
+}
+
+/**
+  Initialize a page table. Only initializes the root page table.
+
+  @retval A pointer to the initialized page table.
+**/
+PAGE_TABLE *
+EFIAPI
+PageTableInit (
+  VOID
+  )
+{
+  UINTN       Pages;
+  PAGE_TABLE  *PageTable;
+
+  Pages = EFI_SIZE_TO_PAGES (sizeof (PAGE_TABLE));
+
+  PageTable = (PAGE_TABLE *)AllocateAlignedPages (Pages, EFI_PAGE_SIZE);
+
+  if (PageTable == NULL) {
+    return NULL;
+  }
+
+  ZeroMem (PageTable, EFI_PAGES_TO_SIZE (Pages));
+
+  return PageTable;
+}
+
+/**
+  Deinitialize a page table.
+
+  @param [in]  Level      The level of the page table to deinitialize.
+  @param [in]  PageTable  The page table to deinitialize.
+**/
+VOID
+EFIAPI
+PageTableDeInit (
+  IN UINT8       Level,
+  IN PAGE_TABLE  *PageTable
+  )
+{
+  UINTN  Index;
+
+  if ((Level >= PAGE_TABLE_DEPTH) || (PageTable == NULL)) {
+    return;
+  }
+
+  for (Index = 0; Index < PAGE_TABLE_SIZE; Index++) {
+    PageTableEntry  Entry = PageTable->Entries[Index];
+
+    if (Entry != 0) {
+      PageTableDeInit (Level + 1, (PAGE_TABLE *)((UINTN)Entry & ~PAGE_TABLE_BLOCK_OFFSET));
+    }
+  }
+
+  FreePages (PageTable, EFI_SIZE_TO_PAGES (sizeof (PAGE_TABLE)));
+}
+
+/**
+  Initialize the IOMMU.
+
+  @retval EFI_SUCCESS  Success.
+**/
+EFI_STATUS
+EFIAPI
+IoMmuInit (
+  VOID
+  )
+{
+  EFI_STATUS  Status;
+  EFI_HANDLE  Handle;
+
+  Handle = NULL;
+  Status = gBS->InstallMultipleProtocolInterfaces (
+                  &Handle,
+                  &gEdkiiIoMmuProtocolGuid,
+                  &SmmuIoMmu,
+                  NULL
+                  );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Failed to install gEdkiiIoMmuProtocolGuid\n", __func__));
+    return Status;

```

Comment:
Not a big deal, but this return is redundant.

Reply:
fixed